### PR TITLE
Arduino_Canvas_Indexed Optimization

### PR DESCRIPTION
**Pull request by jgabriel98**
This pull request have two optimizations:

 - Lowers time complexity of `Arduino_Canvas_Indexed ::get_color_index()` from O(n) to O(1), by using a hashmap (`std::unordered_map`) to keep track of color to index, instead of searching the whole `_color_index[]` array.
 now it should have  a nice speedup.

 - Lowers time complexity of `Arduino_Canvas_Indexed::raise_mask_level()` from O(n*m) to O(m) - where **n** is `COLOR_IDX_SIZE` and **m** is `_framebuffer` size- .
 now it should have a significant speedup
 
 
#### note: if you wish to, you can use only one of them, with some minor changes in the code.

### CODE REVIEW

**Comment by moononournation** [comment]
What is the size of hashmap after draw an JPEG image?

File: src/canvas/Arduino_Canvas_Indexed.cpp [hunk 1]
```diff
@@ -160,22 +162,27 @@ void Arduino_Canvas_Indexed::raise_mask_level()
...
-            for (int32_t i = 0; i < buffer_size; i++)
-            {
-                if (_framebuffer[i] == old_color)
-                {
-                    _framebuffer[i] = new_color;
-                }
-            }
+            uint16_t old_color_indexing = _color_index[old_color];
+            
+            _color_to_index_map.erase(old_color_indexing);      //remove the color with old mask
```
Comment (jgabriel98) [inline src/canvas/Arduino_Canvas_Indexed.cpp:175]: This code assumes that there aren't  `old_color` with the same value of `new_color`.

that is, when the _old colors_ are converted to the _new colors_, there will not exist a value of _new colors_ that matches any value of the _old colors_

**Comment by jgabriel98** [comment]
> What is the size of hashmap after draw an JPEG image?

Unfortunately i don't have a arduino, neither am able to compile the code and test it.
I've just stumbled on this repository and found some possible optimizations. 

it is very probable that this pull request will still need minor fixes (wrong typos, missing commas, etc), and definitely testing.

I apologize for not beeing able to do those essential steps for a pull request. But i've still think that this pull request can bring improvements, after some testing and fixing by a maintener with the required development environment (arduino, C/C++ compiler, libraries)

**Comment by jgabriel98** [comment]
> What is the size of hashmap after draw an JPEG image?

But if i would give an estimative, it's maximum size is arround COLOR_IDX_SIZE  (256 bytes)

The starting size of the hashmap is the same of _color_index, and it will grow only if the _load factor_ is surpassed (explained [here](http://www.cplusplus.com/reference/unordered_map/unordered_map/load_factor/)), but since the default load factor is _1.0_, it will never be surpassed, because there will never be more than `COLOR_IDX_SIZE` keys.

**Comment by moononournation** [comment]
Just info me if you tested the program size can fit in Arduino and can run in MCU limited memory.

**Comment by jgabriel98** [comment]
> Just info me if you tested the program size can fit in Arduino and can run in MCU limited memory.

No i did not, and i cannot test it.
i don't have the necessary tools for it.

**Comment by moononournation** [comment]
rejected untested code.

**Pull request closed without merging**